### ISSUE TITLE
Slow connections: Less Parallelism

### ISSUE THREAD

**Issue by rienesl** [owncloud/client#3382]
### Expected behaviour

Sync multiple big files
### Actual behaviour

One file is aborted with "Connection closed", two others with "Operation canceled"
### Steps to reproduce
1. Have minimum 3 big files (>=100MB) which are not synchronized
2. Have a slow connection (<=56k)
3. Try to sync that
### Server configuration

Operating system:
Linux (Xubuntu)

Web server: 
apache 2.2.22-1

Database:
mysql 5.5.43-0

PHP version:
5.4.42-1

ownCloud version:
8.0.4

Storage backend:
DAS
### Client configuration

Client version:
1.8.1, 1.8.2, 1.8.3

Operating system:
Linux (Xubuntu)

OS language:
English

Installation path of client:
default
### Additional Comments

I guess, everything would run (slow but) fine, if I could deactivate the simultaneous downloads. If you download more than one big file at the time, one of the files will get a timeout because of the slow connection. I have similar problems when downloading multiple files with a browser. There are two ways of solving that problem:
1) Download just one file
2) using a download manager with resume-function
### Edit:

Maybe it has something in common with #3318? Looks similar, but neither this is a Mac system, nor a limit is set

**Comment by guruz** [comment]
Do you mean uploading from desktop or downloading to desktop?
What is your Qt version? (see settings dialog)
Does it work if you ```export OWNCLOUD_MAX_PARALLEL=1``` in the shell before running ```owncloud``` itself?

**Comment by guruz** [comment]
( @ogoffart @dragotin If the ```export``` workaround works: We could tone down parallelism if we detect a slow connection..?)

**Comment by rienesl** [comment]
@guruz:
About the question of up- or downloading: I realized this problem while downloading, I havn't tried it with uploading (seen from desktop client).
About the Qt version and the parameter to set: Currently I'm at work, but I'll check this evening.

**Comment by rienesl** [comment]
Hi, Qt is version 4.8.6. I have started owncloud with the parameter mentioned above, sync started. I'll leave the computer on and go to sleep now. Tomorrow morning, we'll see the results :)

**Comment by rienesl** [comment]
First results seem to be great. This night it successfully downloaded a 103MB file, when I left home it was at 66% of a 146MB file. I'll report this evening, when I'm back home, but I think this parameter does what I need.
About the idea with the detection of a slow connection: If it turns out to work, for me, it would be absolutely sufficient to have a checkbox somewhere in the options: Disable simultaneous downloads

**Comment by rienesl** [comment]
OK, I think I can say, using this parameter works perfect! I was able to synchronize a dozen of files, one of them was 416MB. OK, it took 19 hours, but it was successful! No other errors, no other problems any more

**Comment by guruz** [comment]
@rienesl Thanks for testing :)

**Comment by dragotin** [comment]
Yes, I am all for dynamically adjusting the parallelism. Maybe start with *no* parallelism and increase if we detect proper network. @ogoffart

**Comment by ogoffart** [comment]
This does not seem to be related with slow connection, but with broken connection.  And we have no way to know it until it breaks.  But normaly it shoul dnot be a problem because the client should recover just fine.

**Comment by rienesl** [comment]
@ogoffart: Sorry, but I have to disagree. I do have a _VERY_ reliable connection here, but it is slow. It is just like in the old times, when you had a modem: When downloading too much at a time, your downloads will fail after a while. Nothing had changed about that in the past 20 years...
And about the recovery of the client: It does recover fine, but it will fail again and again and again. I tried to sync those files for over a week now, no success. Then I got the parameter mentioned above and voila: It worked immediately!

**Comment by guruz** [comment]
(btw, I was also thinking of a new checkbox in settings.. automatic can be done later)
As far as i heard, bufferbloat is the cause why reducing paralleism worked..

**Comment by ogoffart** [comment]
I still do not understand why parallelism makes things slower and not working. When you download 3 files at the same time, every file should just take three times more time. So 3 / 3 = 1 and we are at the same speed as if we go not parallel.  But we win a bit because we can continue using the network while the server is working.

Bufferbloat is all about latency. I'm curious to know why it breaks parallelism

**Comment by rienesl** [comment]
@ogoffart: It's simple. You wrote about theory (Quote):"When you download 3 files at the same time, every file should just take three times more time. So 3 / 3 = 1 and we are at the same speed as if we go not parallel."
In reallity, one of the downloads gets prioritized. This choice is not made by client or server or god himself. Most of the time, the download which had been started first will win. This download will use the most bandwidth, the other downloads run in timeouts while waiting for their data, then send a rerequest. This will succeed a time, then the download will fail because of lack of returning data.
What happens in Owncloud: The Server gets a request for 3 files. He answers those request and the same thing described above happens. The funny thing is: 2 downloads fail because of timeouts, then the server closes the connection. But somehow the client is able to finish the "special" one! (yes, it really downloads till complete, I've seen that many times). From this point it is speculative, because I haven't checked that with wireshark: I guess, when it is completed, a final request is send to the server to get a confirmation or whatever. But the server already closed the connection and so instead of the last ack, the client receives an error. <- just as I said, highly speculative!!!

**Comment by dragotin** [comment]
@guruz @ogoffart yes, a checkbox to disable parallelism sounds good IMHO for now.

**Comment by ogoffart** [comment]
I don't think @jancborchardt would like another option for this.
It is already configurable with `OWNCLOUD_MAX_PARALLEL` environement variable for the advanced user.  Maybe we should just document that somewhere.

**Comment by jancborchardt** [comment]
Yeah, this really seems like something which should be fixed in code rather than a setting? And if anything, an advanced setting as @ogoffart says.

**Comment by guruz** [comment]
We decided that we will have some magic when the existing _limit automatically_ radio button is checked (possibly also for the fixed number _limit to x kb/s_)

**Comment by rienesl** [comment]
Result of a long term test: It works perfectly for weeks now. No further errors anymore!
By the way: I read in #3477 (quote): You should be able to put it in ~/.profile
This information makes me _VERY_ happy, because now I can have owncloud in autostart again without starting it from the console every time.

**Comment by guruz** [comment]
https://github.com/facebook/network-connection-class

**Comment by guruz** [comment]
Discussed with @danimo 
We're moving this out of 2.2 because we think this is a server setup or NAT router issue.

The up-scaling of request parallelism (https://github.com/owncloud/client/pull/4529) to improve sync performance will still be done for 2.2

**Comment by guruz** [comment]
@rienesl @dragotin on https://github.com/owncloud/client/pull/4529#issuecomment-191396783

> Yes, of course, it would speed up for small files, but there are still timeout problems when using parallelism with big files on slow connections, so it would be really nice, if you could let decide the user, how much parallelism he want's (dropdownbox?)!

The thing is: Right now only you reported an issue. And for you the issue is fixed by setting bandwidth limiting to "automatic" :+1: 
So to have reasons to further improve this, we need to wait for other users to complain here (or pay money to ownCloud Inc. :-) )

**Comment by rienesl** [comment]
It's pretty easy to simulate:
Wait until you have used off all your included GB in your mobile phone contract (you are in Germany, right?). Take 3-5 iso-images > 700MB and then try to sync.
PS: If the automatic setting will work in combination with more paralellism, I wouldn't mind, but currently (2.1) it is not (and the parameter is not working as well, what forces me to use an old 1.x version)

**Comment by guruz** [comment]
Oh, I thought given https://github.com/owncloud/client/issues/3382#issuecomment-133635698 for you all was fine (with the env variable).. the "automatic" setting was basically the same as setting the environment variable to 1.

Can you go into more detail on what is happening? I don't want to use my mobile quota, but might try later with OS X' Network Link Conditioner

**Comment by rienesl** [comment]
@guruz: Yes setting the environment variable did work perfect. I was able to sync every file I want with no further problems. Then version 2.0 came and this version was a shock. So I reverted to 1.8.4. I waited until the release of 2.1 and gave it a second try: close to everything had improved and was working now, but the automatic setting did not prevent parallelism. Setting the parameter via console did not seem to have any impact either. So again, I reverted back to 1.8.4. To be honest: I did not extensive testing. Maybe something else was or is wrong, but currently I'm on a world trip and I'm relaying a bit on my notebook and data. So screwing something up with experiments is currently not top priority. For now, I'm running on 1.8.4 with the parameter on demand (ie. guesthouses with slow connections) and everything syncs fine. Maybe I overreacted a bit, when I got the mail, that parallelism is going to be increased ;)
PS: No, you shouldn't force off you monthly quota for the test! I meant, that at the end of a month, when it is likely that it is already used up :)
